File: src-tauri/src/types/claude.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct ClaudeHistoryEntry {
    pub display: Option<String>,
    pub timestamp: Option<f64>,
    pub project: Option<String>,
    #[serde(rename = "sessionId")]
    pub session_id: Option<String>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct SessionsIndexFile {
    pub version: Option<i64>,
    pub entries: Option<Vec<SessionsIndexEntry>>,
}

#[derive(Debug, Deserialize, Clone)]
#[allow(dead_code)]
pub struct SessionsIndexEntry {
    #[serde(rename = "sessionId")]
    pub session_id: String,
    #[serde(rename = "projectPath")]
    pub project_path: Option<String>,
    #[serde(rename = "fullPath")]
    pub full_path: Option<String>,
    #[serde(rename = "fileMtime")]
    pub file_mtime: Option<i64>,
    #[serde(rename = "firstPrompt")]
    pub first_prompt: Option<String>,
    pub summary: Option<String>,
    #[serde(rename = "messageCount")]
    pub message_count: Option<i64>,
    pub created: Option<String>,
    pub modified: Option<String>,
    #[serde(rename = "gitBranch")]
    pub git_branch: Option<String>,
    #[serde(rename = "isSidechain")]
    pub is_sidechain: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct ActiveSessionFile {
    pub pid: Option<i64>,
    #[serde(rename = "sessionId")]
    pub session_id: Option<String>,
    pub status: Option<String>,
    #[serde(rename = "updatedAt")]
    pub updated_at: Option<i64>,
    pub kind: Option<String>,
    pub name: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SubagentMeta {
    #[serde(rename = "agentType")]
    pub agent_type: Option<String>,
    pub description: Option<String>,
    pub slug: Option<String>,
}

/// First-line shape of an `agent-*.jsonl` written by older Claude Code
/// (≤ 2.1.74), used as a fallback when no `.meta.json` sidecar exists.
#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct SubagentJsonlHeader {
    #[serde(rename = "agentId")]
    pub agent_id: Option<String>,
    pub slug: Option<String>,
    pub version: Option<String>,
    pub cwd: Option<String>,
}

// Content block types
#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(tag = "type")]
pub enum ContentBlock {
    #[serde(rename = "thinking")]
    Thinking { thinking: String },
    #[serde(rename = "text")]
    Text { text: String },
    #[serde(rename = "tool_use")]
    ToolUse {
        id: Option<String>,
        name: String,
        input: serde_json::Value,
    },
    #[serde(rename = "tool_result")]
    ToolResult {
        tool_use_id: Option<String>,
        content: Option<serde_json::Value>,
        is_error: Option<bool>,
    },
}

#[derive(Debug, Deserialize)]
pub struct TokenUsage {
    pub input_tokens: Option<i64>,
    pub output_tokens: Option<i64>,
    pub cache_creation_input_tokens: Option<i64>,
    pub cache_read_input_tokens: Option<i64>,
    pub inference_geo: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct AssistantMessageInner {
    pub model: Option<String>,
    pub id: Option<String>,
    pub content: Option<Vec<ContentBlock>>,
    pub stop_reason: Option<String>,
    pub usage: Option<TokenUsage>,
}

// We parse each JSONL line as a GenericMessage first,
// then match on the `type` field.
#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct GenericMessage {
    #[serde(rename = "type")]
    pub msg_type: String,

    // Common fields
    pub uuid: Option<String>,
    #[serde(rename = "parentUuid")]
    pub parent_uuid: Option<String>,
    #[serde(rename = "isSidechain")]
    pub is_sidechain: Option<bool>,
    pub timestamp: Option<String>,
    #[serde(rename = "sessionId")]
    pub session_id: Option<String>,
    #[serde(rename = "agentId")]
    pub agent_id: Option<String>,
    #[serde(rename = "isMeta")]
    pub is_meta: Option<bool>,

    // User-specific fields
    #[serde(rename = "gitBranch")]
    pub git_branch: Option<String>,
    pub cwd: Option<String>,
    pub version: Option<String>,
    #[serde(rename = "permissionMode")]
    pub permission_mode: Option<String>,

    // Assistant-specific: nested message object
    pub message: Option<serde_json::Value>,

    // custom-title
    #[serde(rename = "customTitle")]
    pub custom_title: Option<String>,

    // agent-name
    #[serde(rename = "agentName")]
    pub agent_name: Option<String>,

    // ai-title
    #[serde(rename = "aiTitle")]
    pub ai_title: Option<String>,

    // user/tool-result linkage and interruption signal
    #[serde(rename = "sourceToolAssistantUUID")]
    pub source_tool_assistant_uuid: Option<String>,
    #[serde(rename = "toolUseResult")]
    pub tool_use_result: Option<serde_json::Value>,
}

File: src/lib/types/claude.ts
// Raw JSONL line types from ~/.claude/ conversation files

export interface ClaudeHistoryEntry {
	display: string;
	timestamp: number; // Unix ms
	project: string; // Absolute path
	pastedContents: Record<string, unknown>;
	sessionId?: string; // UUID, present in newer entries
}

export interface SessionsIndexFile {
	version: number;
	entries: SessionsIndexEntry[];
}

export interface SessionsIndexEntry {
	sessionId: string;
	fullPath: string;
	fileMtime: number;
	firstPrompt: string;
	messageCount: number;
	created: string; // ISO-8601
	modified: string; // ISO-8601
	gitBranch: string;
	projectPath: string;
	isSidechain: boolean;
}

export interface ActiveSessionFile {
	pid: number;
	sessionId: string;
	cwd: string;
	startedAt: number; // Unix ms
}

export interface SubagentMeta {
	agentType: string;
	description?: string;
}

// Message types in conversation JSONL files

export interface ThinkingBlock {
	type: 'thinking';
	thinking: string;
	signature: string;
}

export interface TextBlock {
	type: 'text';
	text: string;
}

export interface ToolUseBlock {
	type: 'tool_use';
	id: string;
	name: string;
	input: Record<string, unknown>;
	caller?: { type: string };
}

export interface ToolResultBlock {
	type: 'tool_result';
	tool_use_id: string;
	content: string;
	is_error?: boolean;
}

export type ContentBlock = ThinkingBlock | TextBlock | ToolUseBlock | ToolResultBlock;

export interface TokenUsage {
	input_tokens: number;
	output_tokens: number;
	cache_creation_input_tokens?: number;
	cache_read_input_tokens?: number;
	cache_creation?: {
		ephemeral_5m_input_tokens: number;
		ephemeral_1h_input_tokens: number;
	};
	service_tier?: string;
	inference_geo?: string;
	server_tool_use?: {
		web_search_requests: number;
		web_fetch_requests: number;
	};
	iterations?: unknown[];
	speed?: string;
}

export interface ClaudeUserMessage {
	type: 'user';
	parentUuid: string | null;
	isSidechain: boolean;
	userType: string;
	cwd: string;
	sessionId: string;
	version: string;
	gitBranch: string;
	message: {
		role: 'user';
		content: string | ToolResultBlock[];
	};
	uuid: string;
	timestamp: string; // ISO-8601
	permissionMode?: string;
	promptId?: string;
	sourceToolAssistantUUID?: string;
	toolUseResult?: {
		stdout: string;
		stderr: string;
		interrupted: boolean;
		isImage: boolean;
		noOutputExpected: boolean;
	};
	agentId?: string;
	isMeta?: boolean;
}

export interface ClaudeAssistantMessage {
	type: 'assistant';
	parentUuid: string;
	isSidechain: boolean;
	message: {
		model: string;
		id: string;
		type: 'message';
		role: 'assistant';
		content: ContentBlock[];
		stop_reason: string | null;
		stop_sequence: string | null;
		usage: TokenUsage;
	};
	requestId: string;
	uuid: string;
	timestamp: string; // ISO-8601
	userType: string;
	cwd: string;
	sessionId: string;
	version: string;
	gitBranch: string;
	agentId?: string;
	permissionMode?: string;
}

export interface ClaudeProgressMessage {
	type: 'progress';
	parentUuid: string;
	isSidechain: boolean;
	data: {
		type?: 'agent_progress';
		message: unknown;
		prompt?: string;
		agentId?: string;
	};
	toolUseID?: string;
	parentToolUseID?: string;
	uuid: string;
	timestamp: string;
	sessionId: string;
	slug?: string;
}

export interface ClaudeSystemMessage {
	type: 'system';
	subtype?: string;
	durationMs?: number;
	content?: string;
	uuid?: string;
	timestamp?: string;
	sessionId?: string;
}

export interface ClaudeLastPromptMessage {
	type: 'last-prompt';
	lastPrompt: string;
	sessionId: string;
}

export interface ClaudeFileHistorySnapshot {
	type: 'file-history-snapshot';
	messageId: string;
	snapshot: {
		messageId: string;
		trackedFileBackups: Record<string, unknown>;
		timestamp: string;
	};
	isSnapshotUpdate: boolean;
}

export interface ClaudeCustomTitleMessage {
	type: 'custom-title';
	customTitle: string;
	sessionId: string;
}

export interface ClaudeAgentNameMessage {
	type: 'agent-name';
	agentName: string;
	sessionId: string;
}

export type ClaudeMessage =
	| ClaudeUserMessage
	| ClaudeAssistantMessage
	| ClaudeProgressMessage
	| ClaudeSystemMessage
	| ClaudeLastPromptMessage
	| ClaudeFileHistorySnapshot
	| ClaudeCustomTitleMessage
	| ClaudeAgentNameMessage;
